File: prototype/aayu-rs/crates/aayu_vm/src/stack.rs
use crate::value::RuntimeValue;
use aayu_diagnostics::VMError;

#[derive(Debug, Default)]
pub struct RuntimeStack {
    values: Vec<RuntimeValue>,
}

impl RuntimeStack {
    pub fn new() -> Self {
        Self { values: Vec::new() }
    }

    pub fn push(&mut self, val: RuntimeValue) {
        self.values.push(val);
    }

    pub fn pop(&mut self) -> Result<RuntimeValue, VMError> {
        self.values.pop().ok_or(VMError::StackUnderflow)
    }

    pub fn peek(&self) -> Result<&RuntimeValue, VMError> {
        self.values.last().ok_or(VMError::StackUnderflow)
    }

    pub fn len(&self) -> usize {
        self.values.len()
    }

    pub fn is_empty(&self) -> bool {
        self.values.is_empty()
    }
}

File: prototype/language/runtime/vm/stack.py
class Stack:
    def __init__(self):
        self._stack = []

    def push(self, value):
        self._stack.append(value)

    def pop(self):
        if not self._stack:
            raise IndexError("pop from empty stack")
        return self._stack.pop()
        
    def peek(self):
        if not self._stack:
            raise IndexError("peek from empty stack")
        return self._stack[-1]
        
    def is_empty(self):
        return len(self._stack) == 0

    def __len__(self):
        return len(self._stack)
